osu.Game/Rulesets/Difficulty/Preprocessing/DifficultyHitObject.cs

```diff
@@ -41,13 +49,21 @@ public class DifficultyHitObject
         /// <param name="hitObject">The <see cref="HitObject"/> which this <see cref="DifficultyHitObject"/> wraps.</param>
         /// <param name="lastObject">The last <see cref="HitObject"/> which occurs before <paramref name="hitObject"/> in the beatmap.</param>
         /// <param name="clockRate">The rate at which the gameplay clock is run at.</param>
-        public DifficultyHitObject(HitObject hitObject, HitObject lastObject, double clockRate)
+        /// <param name="objects">The list of <see cref="DifficultyHitObject"/>s in the current beatmap.</param>
+        /// <param name="position">The position of this <see cref="DifficultyHitObject"/> in the <see cref="difficultyHitObjects"/> list.</param>
+        public DifficultyHitObject(HitObject hitObject, HitObject lastObject, double clockRate, List<DifficultyHitObject> objects, int position)
```

Comment:
With regards to `lastObject`: I did indeed try this however it ended up being impossible because of the way some ruleset's difficulty calculators skip the first 1 or 2 objects: the very first `DifficultyHitObject` can have a `lastObject` whilst being the first object in the list, so `objects[position - 1` means.... yeah. and I'm not looking to change performances values with the changes here.

As for `hitObject`, yeah I don't see why not

Reply:
Actually, I do see why not: the objects list is a list of `DifficultyHitObject`s, and there's no way to cast them to a `HitObject` as far as I can tell. You can only retrieve their `BaseObject` which is passed through the constructor here. Seems like we need both here

Reply:
I see. Fair enough then I guess, although it's a touch hideous. Do you foresee this going away in some capacity in the future? Maybe even via removing difficulty hitobjects altogether? (bit radical but maybe?)

Reply:
Yeah I could also do with this not overstaying it's welcome. Probably not by removing them since I find them to be pretty valuable, so I'll look into options at a later point (being able to upcast would solve this entirely I think)

Reply:
This is somewhat similar to what I've said in https://github.com/ppy/osu/pull/18057#discussion_r863302374, and figured it can't be removed for the time being due to the ruleset-specific way of not 1:1 mapping hitobjects to `DifficultyHitObject`s.